### PR TITLE
Add XML parser

### PR DESCRIPTION
**Pull request by DmitrySvetlichny**

### CODE REVIEW

File: Sources/Controllers/Help/HelpDetailsViewController.swift [hunk 1]
```diff
@@ -55,6 +74,15 @@ final class HelpDetailsViewController: OABaseNavbarViewController {
                let url = URL(string: urlString) {
                 UIApplication.shared.open(url, options: [:], completionHandler: nil)
             }
+        } else if item.key == "childArticle" {
+            if let childArticles = item.obj(forKey: "childArticles") as? [ArticleNode], !childArticles.isEmpty, let title = item.title {
+                let vc = HelpDetailsViewController(childArticles: childArticles, title: title)
+                self.navigationController?.pushViewController(vc, animated: true)
+            } else if let urlString = item.obj(forKey: "url") as? String {
```
Comment (tigrim) [inline Sources/Controllers/Help/HelpDetailsViewController.swift:81]: add `let item.title`

File: Sources/Controllers/Help/HelpDetailsViewController.swift [hunk 2]
```diff
@@ -12,14 +12,31 @@ import UIKit
...
 final class HelpDetailsViewController: OABaseNavbarViewController {
     var telegramChats: [TelegramChat] = []
+    var childArticles: [ArticleNode] = []
+    var titleText: String?
+    
+    init(childArticles: [ArticleNode], title: String) {
+        self.childArticles = childArticles
+        self.titleText = title
+        super.init(nibName: "OABaseNavbarViewController", bundle: nil)
+        initTableData()
```
Comment (tigrim) [inline Sources/Controllers/Help/HelpDetailsViewController.swift:22]: it's necessary?

File: Sources/Controllers/Help/MenuHelpDataService.swift [hunk 3]
```diff
@@ -127,6 +145,84 @@ final class MenuHelpDataService: NSObject {
...
+            })
+        } else {
+            guard let url = URL(string: urlPrefix + "/sitemap.xml") else {
+                DispatchQueue.main.async {
+                    completion(nil, NSError(domain: "URLCreationError", code: 0, userInfo: nil))
+                }
+                return
+            }
+            
+            URLSession.shared.dataTask(with: url) { [weak self] data, _, error in
```
Comment (tigrim) [inline Sources/Controllers/Help/MenuHelpDataService.swift:172]: [weak self] - u may remove. it's a singleton

File: Sources/Controllers/Help/MenuHelpDataService.swift [hunk 4]
```diff
@@ -136,4 +232,168 @@ final class MenuHelpDataService: NSObject {
...
+            completion(articles, nil)
+        } else {
+            completion(nil, NSError(domain: "XMLParsingError", code: 3, userInfo: nil))
+        }
+    }
+    
+    private func getArticlePropertyName(from url: String) -> String {
+        var propertyName = url.lowercased().replacingOccurrences(of: kOsmAndUserBaseURL, with: "")
+        propertyName = propertyName.replacingOccurrences(of: "-", with: "_").replacingOccurrences(of: "/", with: "_")
+        if !propertyName.isEmpty && propertyName.last == "_" {
```
Comment (tigrim) [inline Sources/Controllers/Help/MenuHelpDataService.swift:270]: && -> ,

File: Sources/Controllers/Help/MenuHelpDataService.swift [hunk 5]
```diff
@@ -136,4 +232,168 @@ final class MenuHelpDataService: NSObject {
...
+            propertyName.removeLast()
+        }
+        
+        return propertyName
+    }
+    
+    private func getSpecialArticleName(for key: String) -> String? {
```
Comment (vshcherb) [inline Sources/Controllers/Help/MenuHelpDataService.swift:277]: Not needed

File: Sources/Controllers/Help/HelpDetailsViewController.swift (hunk 2 above)
Comment (DmitrySvetlichny) [reply Sources/Controllers/Help/HelpDetailsViewController.swift:22]: Yes

File: Sources/Controllers/Help/HelpDetailsViewController.swift [hunk 6]
```diff
@@ -76,10 +76,11 @@ final class HelpDetailsViewController: OABaseNavbarViewController {
...
             let title = item.title
-            if let childArticles = item.obj(forKey: "childArticles") as? [ArticleNode], !childArticles.isEmpty {
-                let vc = HelpDetailsViewController(childArticles: childArticles, title: title ?? "")
+            if let childArticles = item.obj(forKey: "childArticles") as? [ArticleNode], !childArticles.isEmpty, let title = item.title {
+                let vc = HelpDetailsViewController(childArticles: childArticles, title: title)
                 self.navigationController?.pushViewController(vc, animated: true)
-            } else if let urlString = item.obj(forKey: "url") as? String, let webView = OAWebViewController(urlAndTitle: urlString, title: title ?? "") {
+            } else if let urlString = item.obj(forKey: "url") as? String, let title = item.title {
+                guard let webView = OAWebViewController(urlAndTitle: urlString, title: title) else { return }
                 self.navigationController?.pushViewController(webView, animated: true)
```
Comment (tigrim) [inline Sources/Controllers/Help/HelpDetailsViewController.swift:84]: remove self

File: Sources/Controllers/Help/HelpDetailsViewController.swift [hunk 7]
```diff
@@ -76,10 +76,11 @@ final class HelpDetailsViewController: OABaseNavbarViewController {
             }
         } else if item.key == "childArticle" {
             let title = item.title
```
Comment (tigrim) [inline Sources/Controllers/Help/HelpDetailsViewController.swift:78]: add guard

**Review by tigrim — APPROVED** [review]

File: Sources/Controllers/Help/OAHelpViewController.mm [hunk 8]
```diff
@@ -102,53 +105,60 @@ - (void)generateData
...
     {
-        OATableRowData *row = [userGuideSection createNewRow];
-        [row setCellType:[OASimpleTableViewCell getCellIdentifier]];
-        [row setKey:@"userGuide"];
-        [row setTitle:item[@"title"]];
-        [row setIconName:@"ic_custom_book_info"];
-        [row setObj:kLinkInternalType forKey:@"linkType"];
-        [row setObj:item[@"url"] forKey:@"url"];
+        NSString *title = [_helpDataManager getArticleNameFrom:articleNode.url];
+        if (![title isEqualToString:OALocalizedString(@"troubleshooting")])
```
Comment (alex-osm) [inline Sources/Controllers/Help/OAHelpViewController.mm:112]:
```
if ([title isEqualToString:OALocalizedString(@"troubleshooting")])
{
    _troubleshootingChildArticles = articleNode.childArticles;
    continue;
}
```

File: Sources/Controllers/Help/OAHelpViewController.mm [hunk 9]
```diff
@@ -102,53 +105,60 @@ - (void)generateData
...
+    for (ArticleNode *childArticle in _troubleshootingChildArticles)
     {
-        OATableRowData *row = [troubleshootingSection createNewRow];
-        [row setCellType:[OASimpleTableViewCell getCellIdentifier]];
-        [row setKey:@"troubleshooting"];
-        [row setTitle:item[@"title"]];
-        [row setIconName:item[@"icon"]];
-        [row setObj:kLinkInternalType forKey:@"linkType"];
-        [row setObj:item[@"url"] forKey:@"url"];
+        if (![childArticle.url isEqualToString:@"troubleshooting/android_auto/"])
```
Comment (alex-osm) [inline Sources/Controllers/Help/OAHelpViewController.mm:139]:
```
if ([childArticle.url isEqualToString:@"troubleshooting/android_auto/"])
    continue;
```

**Pull request merged**